ssr/src/component/infinite_scroller.rs

```diff
@@ -15,6 +15,16 @@ pub trait KeyedData {
     fn key(&self) -> Self::Key;
 }
 
+pub(crate) trait KeyedCursoredDataProvider<T>:
+    crate::component::infinite_scroller::CursoredDataProvider
+{
+    async fn get_by_cursor_by_key(
+        &self,
+        start: usize,
+        end: usize,
+        user: T,
+    ) -> Result<PageEntry<Self::Data>, Self::Error>;
+}
```

Comment:
As opposed to an entirely new trait, `TokenRootList` can become
```
struct TokenRootList {
    canisters: Canisters<false>,
    user_canister: Principal
}
```

Then you can just use `canisters.get_individual_canister_by_user_principal(self.user_canister)` inside the `CursoredDataProvider` implementation
You can update the history provider implementation similarly